File: packages/runar-go/sdk_wallet_client_integration_test.go
package runar

// ---------------------------------------------------------------------------
// sdk_wallet_client_integration_test.go — live BRC-100 WalletClient round-trip.
//
// Mirrors integration/ruby/spec/wallet_client_spec.rb. Environment-gated:
// runs only when RUNAR_WALLET_ENDPOINT is set to the base URL of a BRC-100
// JSON-over-HTTP wallet endpoint. When unset, the test is skipped cleanly
// so local + CI runs stay green without any wallet setup.
//
// Optional env:
//   RUNAR_WALLET_ENDPOINT — base URL, required
//   RUNAR_WALLET_AUTH     — bearer token, optional
//   RUNAR_WALLET_BASKET   — basket name, default "runar-integration-test"
//
// Asserts the same shape Ruby asserts:
//   * GetPublicKey returns a 33-byte compressed pubkey (66 hex chars).
//   * ListOutputs returns an array; entries (if any) expose at least one of
//     outpoint / satoshis / locking_script.
// ---------------------------------------------------------------------------

import (
	"bytes"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"regexp"
	"strings"
	"testing"
	"time"
)

// httpWalletClient is a tiny BRC-100 JSON-over-HTTP wallet adapter used only
// by this integration test. Speaks POST {endpoint}/{method} with a JSON body
// matching the BRC-100 request shape, returns the parsed JSON response.
type httpWalletClient struct {
	endpoint  string
	authToken string
	client    *http.Client
}

func newHTTPWalletClient(endpoint, authToken string) *httpWalletClient {
	return &httpWalletClient{
		endpoint:  strings.TrimRight(endpoint, "/"),
		authToken: authToken,
		client:    &http.Client{Timeout: 30 * time.Second},
	}
}

func (c *httpWalletClient) post(method string, body any, dst any) error {
	bs, err := json.Marshal(body)
	if err != nil {
		return fmt.Errorf("marshal %s: %w", method, err)
	}
	req, err := http.NewRequest("POST", c.endpoint+"/"+method, bytes.NewReader(bs))
	if err != nil {
		return fmt.Errorf("new request %s: %w", method, err)
	}
	req.Header.Set("Content-Type", "application/json")
	if c.authToken != "" {
		req.Header.Set("Authorization", "Bearer "+c.authToken)
	}
	resp, err := c.client.Do(req)
	if err != nil {
		return fmt.Errorf("do %s: %w", method, err)
	}
	defer resp.Body.Close()
	respBody, _ := io.ReadAll(resp.Body)
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("wallet %s HTTP %d: %s", method, resp.StatusCode, string(respBody))
	}
	if dst == nil {
		return nil
	}
	if err := json.Unmarshal(respBody, dst); err != nil {
		return fmt.Errorf("unmarshal %s: %w", method, err)
	}
	return nil
}

// GetPublicKey calls the BRC-100 getPublicKey endpoint and returns the
// hex-encoded compressed pubkey.
func (c *httpWalletClient) GetPublicKey(protocolID [2]interface{}, keyID string) (string, error) {
	req := map[string]interface{}{
		"protocolID": protocolID,
		"keyID":      keyID,
	}
	var resp struct {
		PublicKey    string `json:"publicKey"`
		PublicKeyHex string `json:"publicKeyHex"`
	}
	if err := c.post("getPublicKey", req, &resp); err != nil {
		return "", err
	}
	if resp.PublicKey != "" {
		return resp.PublicKey, nil
	}
	if resp.PublicKeyHex != "" {
		return resp.PublicKeyHex, nil
	}
	return "", fmt.Errorf("getPublicKey: missing publicKey in response")
}

// ListOutputs calls the BRC-100 listOutputs endpoint and decodes the
// outputs array.
func (c *httpWalletClient) ListOutputs(basket string, tags []string, limit int) ([]map[string]interface{}, error) {
	if tags == nil {
		tags = []string{}
	}
	req := map[string]interface{}{
		"basket": basket,
		"tags":   tags,
		"limit":  limit,
	}
	var resp struct {
		Outputs []map[string]interface{} `json:"outputs"`
	}
	if err := c.post("listOutputs", req, &resp); err != nil {
		return nil, err
	}
	return resp.Outputs, nil
}

func TestWalletClient_LiveEndpoint_RoundTrip(t *testing.T) {
	endpoint := os.Getenv("RUNAR_WALLET_ENDPOINT")
	if endpoint == "" {
		t.Skip("RUNAR_WALLET_ENDPOINT not set — skipping live BRC-100 wallet round-trip. " +
			"Set RUNAR_WALLET_ENDPOINT to a BRC-100 wallet URL to enable.")
	}
	authToken := os.Getenv("RUNAR_WALLET_AUTH")
	basket := os.Getenv("RUNAR_WALLET_BASKET")
	if basket == "" {
		basket = "runar-integration-test"
	}

	wallet := newHTTPWalletClient(endpoint, authToken)
	protocolID := [2]interface{}{2, "runar integration"}
	keyID := "1"

	// 1. getPublicKey: must return a 33-byte compressed secp256k1 key.
	pubKey, err := wallet.GetPublicKey(protocolID, keyID)
	if err != nil {
		t.Fatalf("getPublicKey: %v", err)
	}
	if pubKey == "" {
		t.Fatal("getPublicKey returned empty string")
	}
	if len(pubKey) != 66 {
		t.Fatalf("getPublicKey: expected 66 hex chars, got %d (%q)", len(pubKey), pubKey)
	}
	prefix := pubKey[:2]
	if prefix != "02" && prefix != "03" {
		t.Fatalf("getPublicKey: expected compressed prefix 02/03, got %q", prefix)
	}
	if !regexp.MustCompile(`^[0-9a-fA-F]+$`).MatchString(pubKey) {
		t.Fatalf("getPublicKey: not a hex string: %q", pubKey)
	}
	if _, err := hex.DecodeString(pubKey); err != nil {
		t.Fatalf("getPublicKey: invalid hex: %v", err)
	}

	// 2. listOutputs: must return an array (possibly empty).
	outputs, err := wallet.ListOutputs(basket, nil, 10)
	if err != nil {
		t.Fatalf("listOutputs: %v", err)
	}
	for i, out := range outputs {
		_, hasOutpoint := out["outpoint"]
		_, hasSatoshis := out["satoshis"]
		_, hasLocking := out["lockingScript"]
		if !hasOutpoint && !hasSatoshis && !hasLocking {
			t.Fatalf("listOutputs[%d]: missing canonical outpoint/satoshis/lockingScript fields: %#v", i, out)
		}
	}
}

File: packages/runar-zig/src/sdk_wallet_client_integration_test.zig
// ---------------------------------------------------------------------------
// sdk_wallet_client_integration_test.zig — live BRC-100 WalletClient round-trip.
//
// Mirrors integration/ruby/spec/wallet_client_spec.rb. Environment-gated:
// runs only when RUNAR_WALLET_ENDPOINT is set to the base URL of a BRC-100
// JSON-over-HTTP wallet endpoint. When unset, the test is skipped cleanly so
// local + CI runs stay green without any wallet setup.
//
// Optional env:
//   RUNAR_WALLET_ENDPOINT — base URL, required to actually run
//   RUNAR_WALLET_AUTH     — bearer token, optional
//   RUNAR_WALLET_BASKET   — basket name, default "runar-integration-test"
//
// Asserts:
//   * getPublicKey returns a 33-byte compressed pubkey (66 hex chars,
//     prefix 02/03).
//   * listOutputs returns an array (possibly empty); each entry exposes at
//     least one of outpoint / satoshis / lockingScript.
//
// Implementation: shell-out to curl through std.process.run, mirroring the
// CurlHttpTransport pattern used elsewhere in the SDK.
// ---------------------------------------------------------------------------

const std = @import("std");
const testing = std.testing;

fn envValue(name: []const u8) ?[]const u8 {
    return std.testing.environ.getPosix(name);
}

const PostError = error{
    Curl,
    Http,
    OutOfMemory,
};

fn curlPost(
    allocator: std.mem.Allocator,
    url: []const u8,
    body: []const u8,
    auth_token: ?[]const u8,
) PostError![]u8 {
    var threaded = std.Io.Threaded.init(allocator, .{});
    defer threaded.deinit();

    var argv = std.ArrayListUnmanaged([]const u8).empty;
    defer argv.deinit(allocator);

    argv.appendSlice(allocator, &.{
        "curl",              "-sS", "-f",
        "--connect-timeout", "10",  "--max-time",   "30",
        "-X",                "POST",
        "-H",                "Content-Type: application/json",
        "-d",                body,
    }) catch return PostError.OutOfMemory;

    var auth_header_buf: ?[]u8 = null;
    defer if (auth_header_buf) |b| allocator.free(b);
    if (auth_token) |tok| {
        const h = std.fmt.allocPrint(allocator, "Authorization: Bearer {s}", .{tok}) catch
            return PostError.OutOfMemory;
        auth_header_buf = h;
        argv.append(allocator, "-H") catch return PostError.OutOfMemory;
        argv.append(allocator, h) catch return PostError.OutOfMemory;
    }

    argv.append(allocator, url) catch return PostError.OutOfMemory;

    const res = std.process.run(allocator, threaded.io(), .{ .argv = argv.items }) catch {
        return PostError.Curl;
    };
    defer allocator.free(res.stderr);
    switch (res.term) {
        .exited => |code| if (code != 0) {
            allocator.free(res.stdout);
            return PostError.Http;
        },
        else => {
            allocator.free(res.stdout);
            return PostError.Curl;
        },
    }
    return res.stdout;
}

fn jsonString(value: std.json.Value, key: []const u8) ?[]const u8 {
    if (value != .object) return null;
    const v = value.object.get(key) orelse return null;
    return switch (v) {
        .string => |s| s,
        else => null,
    };
}

test "BRC-100 WalletClient live endpoint round-trip" {
    const endpoint = envValue("RUNAR_WALLET_ENDPOINT") orelse return error.SkipZigTest;
    if (endpoint.len == 0) return error.SkipZigTest;

    const allocator = testing.allocator;
    const auth_token = envValue("RUNAR_WALLET_AUTH");
    const basket = envValue("RUNAR_WALLET_BASKET") orelse "runar-integration-test";

    var trimmed_end = endpoint.len;
    while (trimmed_end > 0 and endpoint[trimmed_end - 1] == '/') : (trimmed_end -= 1) {}
    const trimmed = endpoint[0..trimmed_end];

    // 1. getPublicKey: must return a 33-byte compressed secp256k1 key.
    {
        const url = try std.fmt.allocPrint(allocator, "{s}/getPublicKey", .{trimmed});
        defer allocator.free(url);
        const body =
            \\{"protocolID":[2,"runar integration"],"keyID":"1"}
        ;
        const resp_bytes = try curlPost(allocator, url, body, auth_token);
        defer allocator.free(resp_bytes);

        var parsed = try std.json.parseFromSlice(std.json.Value, allocator, resp_bytes, .{});
        defer parsed.deinit();
        const pub_key = jsonString(parsed.value, "publicKey") orelse
            jsonString(parsed.value, "publicKeyHex") orelse {
            std.debug.print("getPublicKey response missing publicKey: {s}\n", .{resp_bytes});
            return error.TestUnexpectedResult;
        };
        try testing.expectEqual(@as(usize, 66), pub_key.len);
        const prefix = pub_key[0..2];
        try testing.expect(std.mem.eql(u8, prefix, "02") or std.mem.eql(u8, prefix, "03"));
        for (pub_key) |c| {
            const is_hex = (c >= '0' and c <= '9') or (c >= 'a' and c <= 'f') or (c >= 'A' and c <= 'F');
            try testing.expect(is_hex);
        }
    }

    // 2. listOutputs: must return an array (possibly empty).
    {
        const url = try std.fmt.allocPrint(allocator, "{s}/listOutputs", .{trimmed});
        defer allocator.free(url);
        const body = try std.fmt.allocPrint(
            allocator,
            "{{\"basket\":\"{s}\",\"tags\":[],\"limit\":10}}",
            .{basket},
        );
        defer allocator.free(body);
        const resp_bytes = try curlPost(allocator, url, body, auth_token);
        defer allocator.free(resp_bytes);

        var parsed = try std.json.parseFromSlice(std.json.Value, allocator, resp_bytes, .{});
        defer parsed.deinit();
        if (parsed.value != .object) return error.TestUnexpectedResult;
        const outputs_val = parsed.value.object.get("outputs") orelse return; // allow missing
        if (outputs_val != .array) return error.TestUnexpectedResult;
        for (outputs_val.array.items) |out| {
            try testing.expect(out == .object);
            const has_outpoint = out.object.get("outpoint") != null;
            const has_satoshis = out.object.get("satoshis") != null;
            const has_locking = out.object.get("lockingScript") != null;
            try testing.expect(has_outpoint or has_satoshis or has_locking);
        }
    }
}
